pdns/pdnsutil.cc

```diff
@@ -4407,96 +4407,255 @@ static int backendLookup(vector<string>& cmds)
-  {"secure-all-zones", {true, secureAllZones}},
-  {"secure-zone", {true, secureZone}},
-  {"set-account", {true, setAccount}},
-  {"set-catalog", {true, setCatalog}},
-  {"set-kind", {true, setKind}},
-  {"set-meta", {true, setMeta}},
-  {"set-nsec3", {true, setNsec3}},
-  {"set-option", {true, setOption}},
-  {"set-options-json", {true, setOptionsJson}},
-  {"set-presigned", {true, setPresigned}},
-  {"set-publish-cdnskey", {true, setPublishCDNSKey}},
-  {"set-publish-cds", {true, setPublishCDs}},
-  {"show-zone", {true, showZone}},
-  {"test-algorithm", {false, testAlgorithm}},
-  {"test-algorithms", {false, testAlgorithms}},
-  {"test-all-zones", {true, testAllZones}},
-  {"test-schema", {true, testSchema}},
-  {"test-speed", {true, testSpeed}},
-  {"test-zone", {true, testZone}},
-  {"unpublish-zone-key", {true, unpublishZoneKey}},
-  {"unset-nsec3", {true, unsetNSec3}},
-  {"unset-presigned", {true, unsetPresigned}},
-  {"unset-publish-cdnskey", {true, unsetPublishCDNSKey}},
-  {"unset-publish-cds", {true, unsetPublishCDs}},
-  {"verify-crypto", {true, verifyCrypto}},
-  {"zonemd-verify-file", {true, zonemdVerifyFile}}
+enum commandGroup {
+  GROUP_AUTOPRIMARY,
+  GROUP_CATALOG,
+  GROUP_META,
+  GROUP_ZONE,
+  GROUP_RRSET,
+  GROUP_DNSSEC,
+  GROUP_CDNSKEY,
+  GROUP_NSEC3,
+  GROUP_TSIGKEY,
+  GROUP_ZONEKEY,
+  GROUP_OTHER,
+  GROUP_LAST,
+  GROUP_FIRST = GROUP_AUTOPRIMARY,
 };
 
-// NOLINTNEXTLINE(readability-function-cognitive-complexity): TODO Clean this function up.
+static const std::array<std::string_view, GROUP_LAST> groupNames{
+  "Autoprimary",
+  "Catalog",
+  "Metadata",
+  "Zone",
+  "RRSet",
+  "DNSSEC",
+  "CDS/CDNSKEY",
+  "NSEC3",
+  "TSIG key",
+  "Zone key",
+  "Other"
+};
+
+struct commandDispatcher {
+  bool requiresInitialization; // need to invoke reportAllTypes() before handler
+  int (*handler)(std::vector<std::string>&);
+  commandGroup group;
+  const std::string_view help;
+};
+
+// clang-format off
+static const std::unordered_map<std::string, commandDispatcher> commands{
+  {"activate-tsig-key", {true, activateTSIGKey, GROUP_TSIGKEY,
+   "activate-tsig-key ZONE NAME {primary|secondary|producer|consumer}\n"
+   "                                   Enable TSIG authenticated AXFR using the key NAME for ZONE"}},
+  {"activate-zone-key", {true, activateZoneKey, GROUP_ZONEKEY,
+   "activate-zone-key ZONE KEY-ID      Activate the key with key id KEY-ID in ZONE"}},
+  {"add-autoprimary", {true, addAutoprimary, GROUP_AUTOPRIMARY,
+   "add-autoprimary IP NAMESERVER [account]\n"
+   "                                   Add a new autoprimary "}},
+  {"add-meta", {true, setMeta, GROUP_META,
+   "add-meta ZONE KIND VALUE           Add zone metadata, this adds to the existing KIND\n"
+   "                   [VALUE ...]"}},
+  {"add-record", {true, addRecord, GROUP_ZONE,
+   "add-record ZONE NAME TYPE [ttl] content\n"
+   "             [content..]           Add one or more records to ZONE"}},
```

Comment:
```suggestion
   "add-meta ZONE KIND VALUE           Add zone metadata, this adds to the existing KIND\n"
   "                   [VALUE ...]"}},
  {"add-record", {true, addRecord, GROUP_ZONE,
   "add-record ZONE NAME TYPE [ttl] content\n"
   "             [content...]           Add one or more records to ZONE"}},
```

I think you want an extra `.` here, you might want a space depending on what you're doing...